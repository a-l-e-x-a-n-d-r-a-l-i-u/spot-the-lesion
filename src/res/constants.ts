const constants = {
  /* Game */
  aiScoreMultiplier: 75,
  animationCubes: 10,
  animationLineWidth: 3,
  animationTime: 3000,
  clickLineWidth: 3,
  clickSize: 5,
  hintLineWidth: 2,
  hintRadius: 100,
  hintRange: 50,
  predictedLineWidth: 3,
  rounds: 10,
  roundTimeInitial: 10000,
  truthLineWidth: 3,
};

export default constants;
