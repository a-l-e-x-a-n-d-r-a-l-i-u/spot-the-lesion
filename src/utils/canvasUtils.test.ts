import "jest-canvas-mock";
import {
  drawCircle,
  drawCross,
  drawRectangle,
  drawStrokedText,
  mapClickToCanvas,
  mapRectangleToCanvasScale,
  toCanvasScale,
  toDefaultScale,
} from "./canvasUtils";
import constants from "../res/constants";

let ctx;

beforeAll(() => {
  const canvas = document.createElement("canvas");

  canvas.width = constants.defaultImageSize / 2;
  canvas.height = constants.defaultImageSize / 2;

  ctx = canvas.getContext("2d");
});

afterEach(() => {
  // eslint-disable-next-line no-underscore-dangle
  ctx.__clearEvents();
});

describe("drawCross", () => {
  it("correctly uses input parameters", () => {
    drawCross(ctx, 0, 0, 5, 3, "#ffa500");

    expect(ctx.lineWidth).toBe(3);
    expect(ctx.strokeStyle).toBe("#ffa500");
    expect(ctx.beginPath).toBeCalled();
    expect(ctx.moveTo).toBeCalledWith(-5, -5);
    expect(ctx.lineTo).toBeCalledWith(5, 5);
    expect(ctx.moveTo).toBeCalledWith(5, -5);
    expect(ctx.lineTo).toBeCalledWith(-5, 5);
    expect(ctx.stroke).toBeCalled();
  });

  it("renders correctly", () => {
    drawCross(ctx, 0, 0, 5, 3, "#ffa500");

    // eslint-disable-next-line no-underscore-dangle
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });
});

describe("drawCircle", () => {
  it("correctly uses input parameters", () => {
    drawCircle(ctx, 0, 0, 5, 3, "#ffa500");

    expect(ctx.lineWidth).toBe(3);
    expect(ctx.strokeStyle).toBe("#ffa500");

    expect(ctx.beginPath).toBeCalled();
    expect(ctx.arc).toBeCalledWith(0, 0, 5, 0, 2 * Math.PI);
    expect(ctx.stroke).toBeCalled();
  });

  it("renders correctly", () => {
    drawCircle(ctx, 0, 0, 5, 3, "#ffa500");

    // eslint-disable-next-line no-underscore-dangle
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });
});

describe("drawRectangle", () => {
  it("correctly uses input parameters", () => {
    drawRectangle(ctx, [0, 0, 5, 5], 3, "#ffa500");

    expect(ctx.lineWidth).toBe(3);
    expect(ctx.strokeStyle).toBe("#ffa500");

    expect(ctx.beginPath).toBeCalled();
    expect(ctx.rect).toBeCalledWith(0, 0, 5, 5);
    expect(ctx.stroke).toBeCalled();
  });

  it("throws error for invalid input rectangle", () => {
    expect(() => drawRectangle(ctx, [], 3, "#ffa500")).toThrow();
    expect(() => drawRectangle(ctx, [1, 3], 3, "#ffa500")).toThrow();
    expect(() => drawRectangle(ctx, [1, 3, 5, 6, 7], 3, "#ffa500")).toThrow();
  });

  it("renders correctly", () => {
    drawRectangle(ctx, [0, 0, 5, 5], 3, "#ffa500");

    // eslint-disable-next-line no-underscore-dangle
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });
});

describe("drawStrokedText", () => {
  it("correctly uses input parameters", () => {
    drawStrokedText(ctx, "text", 0, 0, "center", 10, 5, "#000000", "#ffffff");

    expect(ctx.font).toBe("10px Roboto");
    expect(ctx.lineJoin).toBe("miter");
    expect(ctx.miterLimit).toBe(2);

    expect(ctx.textAlign).toBe("center");
    expect(ctx.lineWidth).toBe(5);
    expect(ctx.strokeStyle).toBe("#000000");
    expect(ctx.fillStyle).toBe("#ffffff");

    expect(ctx.strokeText).toBeCalledWith("text", 0, 0);
    expect(ctx.fillText).toBeCalledWith("text", 0, 0);
  });

  it("renders correctly", () => {
    drawStrokedText(ctx, "text", 0, 0, "center", 10, 5, "#000000", "#ffffff");

    // eslint-disable-next-line no-underscore-dangle
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });
});

describe("toDefaultScale", () => {
  it("correctly maps to default scale", () => {
    expect(toDefaultScale(ctx, 100)).toBe(200);
  });

  it("is inverse of toCanvasScale", () => {
    expect(toDefaultScale(ctx, toCanvasScale(ctx, 200))).toBe(200);
  });
});

describe("toCanvasScale", () => {
  it("correctly maps to canvas scale", () => {
    expect(toCanvasScale(ctx, 200)).toBe(100);
  });

  it("is inverse of toDefaultScale", () => {
    expect(toCanvasScale(ctx, toDefaultScale(ctx, 200))).toBe(200);
  });
});

describe("mapRectangleToCanvasScale", () => {
  it("is equivalent to calling toCanvasScale on every coordinate", () => {
    const mappedRect = mapRectangleToCanvasScale(ctx, [1, 3, 5, 7]);

    const mapped1 = toCanvasScale(ctx, 1);
    const mapped3 = toCanvasScale(ctx, 3);
    const mapped5 = toCanvasScale(ctx, 5);
    const mapped7 = toCanvasScale(ctx, 7);

    expect(mappedRect).toStrictEqual([mapped1, mapped3, mapped5, mapped7]);
  });
});

describe("mapClickToCanvas", () => {
  let spyGetBoundingClientRect;

  beforeAll(() => {
    spyGetBoundingClientRect = jest.spyOn(Element.prototype, "getBoundingClientRect");
  });

  afterAll(() => spyGetBoundingClientRect.mockRestore());

  afterEach(() => spyGetBoundingClientRect.mockClear());

  it("correctly maps to canvas", () => {
    spyGetBoundingClientRect.mockReturnValue({
      height: ctx.canvas.height * 2,
      width: ctx.canvas.width * 2,
      left: 50,
      top: 50,
    });

    const mappedClick = mapClickToCanvas(ctx, 100, 200);

    expect(spyGetBoundingClientRect).toBeCalled();

    expect(mappedClick).toStrictEqual({ x: 25, y: 75 });
  });
});
