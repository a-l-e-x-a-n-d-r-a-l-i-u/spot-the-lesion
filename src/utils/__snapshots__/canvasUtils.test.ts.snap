// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawCircle renders correctly 1`] = `
Array [
  Object {
    "props": Object {
      "value": 3,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "#ffa500",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
]
`;

exports[`drawCross renders correctly 1`] = `
Array [
  Object {
    "props": Object {
      "value": 3,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "#ffa500",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
]
`;

exports[`drawRectangle renders correctly 1`] = `
Array [
  Object {
    "props": Object {
      "value": 3,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "#ffa500",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
]
`;

exports[`drawStrokedText renders correctly 1`] = `
Array [
  Object {
    "props": Object {
      "value": "10px Roboto",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "font",
  },
  Object {
    "props": Object {
      "value": "miter",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineJoin",
  },
  Object {
    "props": Object {
      "value": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "center",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "textAlign",
  },
  Object {
    "props": Object {
      "value": 5,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "lineWidth",
  },
  Object {
    "props": Object {
      "value": "#000000",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "strokeStyle",
  },
  Object {
    "props": Object {
      "value": "#ffffff",
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillStyle",
  },
]
`;
